url70ed135da7_ralfbiedert_jcores_tgz-pJ8-sandbox_SimpleFilterTestJ8/src/net/jcores/jre/cores/CoreURI.java

```diff
@@ -0,0 +1,235 @@
+ * 
+ * <code>$(uri).download("downloads")</code>
+ * 
+ * @author Ralf Biedert
+ * 
+ * @since 1.0
+ */
+public class CoreURI extends CoreObject<URI> {
+
+    /** Used for serialization */
+    private static final long serialVersionUID = 7366734773387957013L;
+
+    /**
+     * Creates an URI core.
+     * 
+     * @param supercore The common core.
+     * @param objects The adapter to wrap.
+     */
+    public CoreURI(CommonCore supercore, URI... objects) {
+        super(supercore, objects);
+    }
+    
+    /**
+     * Creates an URI core.
+     * 
+     * @param supercore The common core.
+     * @param adapter The adapter to wrap.
+     */
+    public CoreURI(CommonCore supercore, AbstractAdapter<URI> adapter) {
+        super(supercore, adapter);
+    }
+
+
+    /**
+     * Opens the associated input stream.<br/>
+     * <br/>
+     * 
+     * Examples:
+     * <ul>
+     * <li><code>$("http://jcores.net/index.html").uri().input()</code> - Opens an input stream for the given URI.</li>
+     * </ul>
+     * 
+     * Multi-threaded.<br/>
+     * <br/>
+     * @param options Optional arguments, especially {@link OnFailure}.
+     * 
+     * @return A CoreInputStream object enclosing the opened input streams.
+     */
+    @SupportsOption(options = {OnFailure.class})
+    public CoreInputStream input(Option ... options) {
+        final Options options$ = Options.$(this.commonCore, options);
+
+        return new CoreInputStream(this.commonCore, map(new F1<URI, InputStream>() {
+            public InputStream f(URI x) {
+                try {
+                    final URL url = x.toURL();
+                    final InputStream openStream = url.openStream();
+                    return openStream;
+                } catch (MalformedURLException e) {
+                    options$.failure(x, e, "input:urimalformed", "Malformed URI.");
+                } catch (IOException e) {
+                    options$.failure(x, e, "input:ioerror", "Error opening the URI.");
+                }
+
+                return null;
+            }
+        }).array(InputStream.class));
+    }
+    
+    /**
+     * Downloads the enclosed URIs to a temporary directories and returns core
+     * containing their filenames.<br/>
+     * <br/>
+     * 
+     * Examples:
+     * <ul>
+     * <li><code>$("http://jcores.net/index.html").uri().download()</code> - Downloads the file at the given URI to the temporary directory.</li>
+     * </ul>
+     * 
+     * Multi-threaded.<br/>
+     * <br/>
+     * 
+     * @return A CoreFile object enclosing the files of all downloaded URIs.
+     */
+    public CoreFile download() {
+        final CommonCore cc = this.commonCore;
+
+        return new CoreFile(this.commonCore, map(new F1<URI, File>() {
+            public File f(URI x) {
+                try {
+                    final URL url = x.toURL();
+                    final InputStream openStream = url.openStream();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *openStream*. The resource is referenced at line 145. The resource is closed at line 147. There are other execution paths that do not contain closure statements, for example, when File.createTempFile throws an exception. Ensure *openStream* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).